docs/control-structures.rst

```diff
@@ -60,31 +63,73 @@ Input parameters and output parameters can be used as expressions in
+
+.. _return-array:
+
+Returning Strings and Arrays
+----------------------------
+
+You can return strings and arrays from Solidity functions, see `array_receiver_and_returner.sol <https://github.com/fivedogit/solidity-baby-steps/blob/master/contracts/60_array_receiver_and_returner.sol>`_ for an example.
+
+What is problematic, though, is returning any variably-sized data (e.g. a variably-sized array like ``uint[]`` or an array of strings ``string[]``) from a function **called from within Solidity**. This is a limitation of the EVM and will be solved with the next protocol update.
+
+Returning variably-sized data as part of an external transaction or call doesn't have this problem.
+
+.. index:: return struct, struct
+
+.. _return-struct:
+
+Returning Structs
+-----------------
+
+You can return a ``struct`` from a Solidity function, but only from ``internal`` function calls.
+
 .. index:: if, else, while, do/while, for, break, continue, return, switch, goto
 
 Control Structures
 ===================
 
-Most of the control structures from JavaScript are available in Solidity
-except for ``switch`` and ``goto``. So
-there is: ``if``, ``else``, ``while``, ``do``, ``for``, ``break``, ``continue``, ``return``, ``? :``, with
-the usual semantics known from C or JavaScript.
+Most of the `control structures from JavaScript <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Control_flow_and_error_handling>`_ are available in Solidity except for ``switch`` and ``goto``. The full list is:
 
-Parentheses can *not* be omitted for conditionals, but curly brances can be omitted
-around single-statement bodies.
+- ``if``
+- ``else``
+- ``while``
+- ``do``
+- ``for``
+- ``break``
+- ``continue``
+- ``return``
+- ``? :``
 
-Note that there is no type conversion from non-boolean to boolean types as
-there is in C and JavaScript, so ``if (1) { ... }`` is *not* valid
-Solidity.
+All with the usual semantics from C or JavaScript, with the following caveats.
 
-.. _multi-return:
+``for`` loops
+-------------
 
-Returning Multiple Values
--------------------------
+If you use syntax like ``for (var i = 0; i < a.length; i ++) { a[i] = i; }``, then Solidity implies the type of ``i`` only from ``0``, whose type is ``uint8``. This means that if ``a`` has more than ``255`` elements, your loop will not terminate because ``i`` can only hold values up to ``255``.
 
-When a function has multiple output parameters, ``return (v0, v1, ...,
-vn)`` can return multiple values.  The number of components must be
-the same as the number of output parameters.
+It's better to use syntax like ``for (uint i = 0; i < a.length; i ++) { a[i] = i; }``, see `struct_and_for_loop_tester.sol <https://github.com/fivedogit/solidity-baby-steps/blob/master/contracts/65_struct_and_for_loop_tester.sol>`_ for an example.
+
+
+Conditional Values
+------------------
+
+You *cannot* omit parentheses for conditionals, but you can omit curly braces around single-statement bodies.
```

Comment:
This is also the case in C and JavaScript, it is only special for people coming from e.g. go.

Reply:
@chriseth So I think in this case, as we specifically mention C and JavaScript above, just remove this section?